adding nvidia models

nv-kkudrynski opened the pull request.

ailzhang commented on scripts/run_pytorch.sh:
```diff
@@ -6,12 +6,14 @@ conda activate base
-  # FIXME: GAN models checkpoints are on cuda.
-  if [[ $f = $GANs* ]]; then
+  # FIXME: GAN and NVIDIA models checkpoints are on cuda.
+  if [[ $f = $GANs* ]] || [[ $f = $CUDAs* ]]; then
```
Adding a cuda checkpoint will require user have a cuda pytorch installed to load it. Would it possible to convert them to cpu in hubconf entrypoints and relax it here?

soumith replied:
considering this is NVIDIA, i think it's okay

nv-kkudrynski replied:
thx. I had to revive the above condition in run_pytorch.sh to pass your CI

ailzhang commented:
@nv-kkudrynski Also I'm not sure why this didn't trigger the circleCi check. See #23 as an example, maybe try rebasing?

ailzhang commented on nvidia_deeplearningexamples_tacotron2.md:
````diff
@@ -0,0 +1,54 @@
+
+```python
+import torch
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_tacotron2')
````
These two args are positional args, you can remove the keywords I think

nv-kkudrynski replied:
ok

ailzhang commented on nvidia_deeplearningexamples_tacotron2.md:
```diff
@@ -0,0 +1,54 @@
+layout: hub_detail
+background-class: hub-background
+body-class: hub
+title: NCF
```
Title doesn't match model ;)

nv-kkudrynski replied:
sure. i already noticed it on netlify. corrected. thx

ailzhang commented on nvidia_deeplearningexamples_waveglow.md:
```diff
@@ -0,0 +1,52 @@
+layout: hub_detail
+background-class: hub-background
+body-class: hub
+title: NCF
```
Title doesn't match model ;)

nv-kkudrynski replied:
sure. i already noticed it on netlify. corrected. thx

gottbrath commented:
@soumith  -- are we going to be able to accept this whole pull request? I believe from side conversation you had some concerns with the NCF. Have those been addressed? 

If we can't accept the NCF part would it make sense to split this so that we could get the waveglow and tacotron2 models up first then address the NCF issues?

soumith commented:
yes, there are a few changes needed.

1. can we remove NCF from this PR for now -- I dont think the model as it stands is useful for Hub (we are curating), because it isn't really useful for building research on top of it -- or useful to visualize anything about the output
2. Can you add code that visualizes the outputs of WaveGlow / TacoTron-2. They produce spectrograms, that's great -- it'd be useful to see that. You can use the GAN model that I cleaned up as an example: https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/ . This way, people pulling your model know what to do with it further.

soumith commented on nvidia_deeplearningexamples_tacotron2.md:
```diff
@@ -0,0 +1,54 @@
+import torch
+print('\nLoading tacotron2 model from torch.hub.')
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_tacotron2')
+hub_model = hub_model.cuda()
```
considering that it's NVIDIA providing the model, i think it's okay to let this stand.

soumith commented on nvidia_deeplearningexamples_tacotron2.md:
```diff
@@ -0,0 +1,54 @@
+hub_model = hub_model.cuda()
+hub_model.eval()
+inp = torch.randint(low=0, high=148, size=(1,140), dtype=torch.long)
+inp = torch.autograd.Variable(inp).cuda().long()
```
yes, please remove `torch.autograd.Variable(`

nv-kkudrynski replied:
sure. thx. done

soumith commented on nvidia_deeplearningexamples_tacotron2.md:
```diff
@@ -0,0 +1,54 @@
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_tacotron2')
+hub_model = hub_model.cuda()
+hub_model.eval()
+inp = torch.randint(low=0, high=148, size=(1,140), dtype=torch.long)
```
describe what this input is supposed to be, for example see https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/

nv-kkudrynski replied:
changed strategy of this example to go from plain text to sound using the two torch.hub models in series

soumith commented on nvidia_deeplearningexamples_waveglow.md:
````diff
@@ -0,0 +1,52 @@
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_waveglow')
+```
+will load the WaveGlow model pre-trained on [LJ Speech dataset](https://keithito.com/LJ-Speech-Dataset/)
+
````
same thoughts as above. Add details on what the model takes as input and gives as output.
For example see: https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/

nv-kkudrynski replied:
changed strategy of this example to go from plain text to sound using the two torch.hub models in series

soumith commented on nvidia_deeplearningexamples_waveglow.md:
````diff
@@ -0,0 +1,52 @@
+print('\nWaveglow model test output:')
+print(out.size())
+```
+
````
add a visualization of the test output, or some other way to interpret the output

nv-kkudrynski replied:
changed strategy of this example to go from plain text to sound using the two torch.hub models in series

nv-kkudrynski commented:
> 1. can we remove NCF from this PR for now -- I dont think the model as it stands is useful for Hub (we are curating), because it isn't really useful for building research on top of it -- or useful to visualize anything about the output

Indeed, NCF is quite specific. We also had discussions about it. It could however be useful if loaded pretrained=False (I added example in readme and provided reference for details about training). Then pre-trained model could serve as reference. Let me know what you think.

> 2. Can you add code that visualizes the outputs of WaveGlow / TacoTron-2. They produce spectrograms, that's great -- it'd be useful to see that. You can use the GAN model that I cleaned up as an example: https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/ . This way, people pulling your model know what to do with it further.

In our flow formatting input/output is not part of the model object, so we need a few internal changes to adapt to your request. We are working on this atm.

soumith commented:
@nv-kkudrynski i think NCF can serve as a reference with `pretrained=False` if it's easy to train upon. Is the code snippet that shows -- taking a csv dataset or Pandas Dataframe, training the NCF model with it -- going to be about ~20 lines or less? If so, I think it can be worth it for the reader who comes to the hub page. If the reference model is only useful if the data is preprocessed and run through a high-performance set of instructions that are given in the NCF repo, then it's almost better to just point users to the NCF repo directly without a card.

soumith commented:
even an "easy-finetuning" starting from the pretrained model possibility has potential, but from what I can tell with the dataset, the pre-trained model doesn't have much to gain wrt fine-tuning -- the features aren't generic to generalize to a different set of users + attributes to do recommendation on

soumith commented:
> In our flow formatting input/output is not part of the model object, so we need a few internal changes to adapt to your request. We are working on this atm.

Great, thanks for the update.
Just FYI, if the input our output is not part of the model object, that is okay. This is why hub entrypoints are callables and not models.
For example, you can have an entry-point for pre-processing, and an entry-point for post-processing.
In the [BERT](https://pytorch.org/hub/huggingface_pytorch-pretrained-bert_bert/) example, they have an entry `bertTokenizer` which is not a model, it's just a pre-processing tokenizer -- but it's available as an entrypoint.
The example showcases taking an input, tokenizing it, and then passing in the output of `bertTokenizer` into `bertModel`. Maybe something like this will help with your case as well, just a thought.

nv-kkudrynski commented:
> Just FYI, if the input our output is not part of the model object, that is okay. This is why hub entrypoints are callables and not models.

I think there was a discussion about it which concluded with exactly the opposite :) 
Which is reflected in your guidelines at: https://pytorch.org/docs/stable/hub.html:
>Entrypoint function should ALWAYS return a model(nn.module).

But don't worry, update is coming ;)

soumith commented:
@ailzhang that line needs to be removed I think?

nv-kkudrynski commented:
I removed NCF from this PR to enable tacotron2/waveglow hopefully before the weekend.
Now the 2 models play very nicely together in the examples - producing sound from input text.

gottbrath commented:
@soumith  - have the changes here addressed all your concerns?

soumith commented:
thanks so much @nv-kkudrynski , the updated stuff looks good. I might make some additional changes so that it plays well with opening the notebooks in Google Colab.

Additionally, if you are interested as a follow-up (totally up to you and if it makes sense), add end-points for any intermediate features that might make sense for folks to use these models as a feature-extractor / fine-tuner.

nv-kkudrynski commented:
Thanks to everyone for all the feedback and smooth cooperation!

soumith commented:
fyi, I have updated the hub card to play well with google colab -- now you can generate the speech and listen to it immediately, all in the colab notebook.

nv-kkudrynski commented:
Thanks a lot for that!
